File: C++/78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */
#include <vector>
#include <algorithm>
using namespace std;


// backtracking
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        
        vector<vector<int>> res = {{}};

        if(nums.size() == 0)
            return res;

        vector<int> temp = {};

        Rec(0,temp, res, nums);

        return res;

    }

    void Rec(int n, vector<int>& temp, vector<vector<int>>& res, vector<int>& nums)
    {
        if(n >= nums.size())
            return;
        else
        {
            temp.push_back(nums[n]);
            res.push_back(temp);
            Rec(n+1, temp, res, nums);
            temp.pop_back();
            Rec(n+1, temp, res, nums);
        }
    }    
};

// // complete with bitwise operation.
// // treat the set as a integer. The first element is represented by b'0001, the second is b'0010
// // then the whole set of subsets is from 1 to 2^n
// // To find if each element exists in the i-th subset, just do the bitwise AND
// // 1<<nums[i] & i
// class Solution {
// public:
//     vector<vector<int>> subsets(vector<int>& nums) {
//         int len = nums.size();
//         if(len == 0)
//             return vector<vector<int>> {{}};
//         vector<vector<int>> res;
//         for(int i = 0; i < (1<<len); i++)
//         {
//             vector<int> temp {};
//             for(int j = 0; j < len; j++)
//             {
//                 if((1<<j)&i)
//                     temp.push_back(nums[j]);
//             }
//             res.push_back(temp);
//         }
//         return res;
//     }    
// };

File: Java/78.subsets.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {

    // time complexity: O(n * 2^n)
    // space complexity: O(2^n)

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, 0, new ArrayList<>(), res);
        return res;
    }

    private void dfs(int[] nums, int index, List<Integer> prefix, List<List<Integer>> res) {
        if (index == nums.length) {
            List<Integer> newSubset = createSubset(prefix);
            res.add(newSubset);
            return;
        }
        // not add current element
        dfs(nums, index + 1, prefix, res);
        
        // add current element
        prefix.add(nums[index]);
        dfs(nums, index + 1, prefix, res);
        prefix.remove(prefix.size() - 1);
    }

    private List<Integer> createSubset(List<Integer> nums) {
        List<Integer> l = new ArrayList<>(nums.size());
        for (int i : nums) {
            l.add(i);
        }
        return l;
    }
}
// @lc code=end